app/src/org/commcare/android/resource/analytics/UpdateStats.java

```diff
@@ -0,0 +1,184 @@
+package org.commcare.android.resource.analytics;
+
+import android.content.SharedPreferences;
+import android.util.Base64;
+import android.util.Log;
+
+import org.commcare.dalvik.application.CommCareApp;
+import org.commcare.resources.model.InstallStatsLogger;
+
+import java.io.ByteArrayInputStream;
+import java.io.ByteArrayOutputStream;
+import java.io.IOException;
+import java.io.ObjectInputStream;
+import java.io.ObjectOutputStream;
+import java.io.PrintWriter;
+import java.io.Serializable;
+import java.io.StringWriter;
+import java.util.Date;
+import java.util.Hashtable;
+
+/**
+ * Statistics associated with attempting to stage resources into the app's
+ * update table.
+ *
+ * @author Phillip Mates (pmates@dimagi.com)
+ */
+public class UpdateStats implements InstallStatsLogger, Serializable {
+    private static final String TAG = UpdateStats.class.getSimpleName();
+    private final Hashtable<String, InstallAttempts<String>> resourceInstallStats;
+    private final long startInstallTime;
+    private int restartCount = 0;
+    private final static String TOP_LEVEL_STATS_KEY = "top-level-update-exceptions";
+    private static final String UPGRADE_STATS_KEY = "upgrade_table_stats";
+
+    private static final long TWO_WEEKS_IN_MS = 1000 * 60 * 60 * 24 * 24;
+    private static final int ATTEMPTS_UNTIL_UPDATE_STALE = 5;
+
+    private UpdateStats() {
+        startInstallTime = new Date().getTime();
+        resourceInstallStats = new Hashtable<>();
+        resourceInstallStats.put(TOP_LEVEL_STATS_KEY,
+                new InstallAttempts<String>(TOP_LEVEL_STATS_KEY));
+    }
+
+    /**
+     * Load update statistics associated with upgrade table from app preferences
+     *
+     * @return Persistently-stored update stats or if no stats found then a new
+     * update stats object.
+     */
+    public static UpdateStats loadUpdateStats(CommCareApp app) {
+        SharedPreferences prefs = app.getAppPreferences();
+        if (prefs.contains(UPGRADE_STATS_KEY)) {
+            try {
+                String serializedObj = prefs.getString(UPGRADE_STATS_KEY, "");
+                return (UpdateStats)UpdateStats.deserialize(serializedObj);
+            } catch (Exception e) {
+                Log.w(TAG, "Failed to deserialize update stats, defaulting to new instance.");
+                e.printStackTrace();
+                clearPersistedStats(app);
+                return new UpdateStats();
+            }
+        } else {
+            return new UpdateStats();
+        }
+    }
+
+    private static Object deserialize(String s) throws IOException,
+            ClassNotFoundException {
+        byte[] data = Base64.decode(s, Base64.DEFAULT);
```

Comment:
Assuming we're treating it as a loss if this changes over time and the serialization breaks?